homeassistant/helpers/entity_component.py

```diff
@@ -126,40 +182,60 @@ def add_entity(self, entity, platform=None):
                 object_id = '{} {}'.format(platform.entity_namespace,
                                            object_id)
 
-            entity.entity_id = generate_entity_id(
+            entity.entity_id = async_generate_entity_id(
                 self.entity_id_format, object_id,
                 self.entities.keys())
 
         self.entities[entity.entity_id] = entity
-        entity.update_ha_state()
+        yield from entity.async_update_ha_state()
 
         return True
 
     def update_group(self):
         """Set up and/or update component group."""
+        run_callback_threadsafe(
+            self.hass.loop, self.async_update_group).result()
+
+    @asyncio.coroutine
+    def async_update_group(self):
+        """Set up and/or update component group.
+
+        This method must be run in the event loop.
+        """
         if self.group is None and self.group_name is not None:
             group = get_component('group')
-            self.group = group.Group(self.hass, self.group_name,
-                                     user_defined=False)
-
-        if self.group is not None:
-            self.group.update_tracked_entity_ids(self.entities.keys())
+            self.group = yield from group.Group.async_create_group(
+                self.hass, self.group_name, self.entities.keys(),
+                user_defined=False
+            )
+        elif self.group is not None:
+            yield from self.group.async_update_tracked_entity_ids(
+                self.entities.keys())
 
     def reset(self):
         """Remove entities and reset the entity component to initial values."""
-        with self.lock:
-            for platform in self._platforms.values():
-                platform.reset()
+        run_coroutine_threadsafe(self.async_reset(), self.hass.loop).result()
 
-            self._platforms = {
-                'core': self._platforms['core']
-            }
-            self.entities = {}
-            self.config = None
+    @asyncio.coroutine
+    def async_reset(self):
+        """Remove entities and reset the entity component to initial values.
 
-            if self.group is not None:
-                self.group.stop()
-                self.group = None
+        This method must be run in the event loop.
+        """
+        tasks = [platform.async_reset() for platform
```

Comment:
@bbangert question for you:

@pvizeli pointed out to me that I have forgotten to wrap my coroutines in `self.hass.loop.create_task(…)` but it still works. After reading up on [the docs](https://docs.python.org/3/library/asyncio-task.html#asyncio.gather) this seems correct: `gather` will run the coroutines and aggregate the results.

Would there be any benefit in wrapping the coroutines in a task object here?